server/server.go

```diff
@@ -451,16 +456,52 @@ func (s *Server) unlimitedDo(pkt *protocol.Packet, connName string) response {
 	signSpan, _ := opentracing.StartSpanFromContext(ctx, "execute.Sign")
 	defer signSpan.Finish()
 	var sig []byte
-	sig, err = key.Sign(rand.Reader, pkt.Operation.Payload, opts)
-	if err != nil {
-		tracing.LogError(span, err)
-		log.Errorf("Connection %v: %s: Signing error: %v\n", connName, protocol.ErrCrypto, err)
-		return makeErrResponse(pkt, protocol.ErrCrypto)
+	for attempts := 1 + s.signRetryCount; attempts > 0; attempts-- {
+		var err error
+		// If signTimeout is not set, the value will be zero
+		if s.signTimeout == 0 {
+			sig, err = key.Sign(rand.Reader, pkt.Operation.Payload, opts)
+		} else {
+			ch := make(chan signWithTimeoutWrapper, 1)
+			ctxTimeout, cancel := context.WithTimeout(context.Background(), s.signTimeout)
+			defer cancel()
+
+			go signWithTimeout(ctxTimeout, ch, key, rand.Reader, pkt.Operation.Payload, opts)
+			select {
+			case <-ctxTimeout.Done():
+				sig = nil
+				err = ctxTimeout.Err()
+			case result := <-ch:
+				sig = result.sig
+				err = result.error
+			}
+		}
+		if err != nil {
+			if attempts > 1 {
+				log.Debugf("Connection %v: failed sign attempt: %s, %d attempt(s) left", connName, err, attempts-1)
+				continue
+			} else {
+				tracing.LogError(span, err)
+				log.Errorf("Connection %v: %s: Signing error: %v\n", connName, protocol.ErrCrypto, err)
+				return makeErrResponse(pkt, protocol.ErrCrypto)
+			}
+		}
+		break
 	}
 
 	return makeRespondResponse(pkt, sig)
 }
 
+type signWithTimeoutWrapper struct {
+	sig   []byte
+	error error
+}
+
+func signWithTimeout(ctx context.Context, ch chan signWithTimeoutWrapper, key crypto.Signer, rand io.Reader, digest []byte, opts crypto.SignerOpts) {
```

Comment:
Maybe a future refactor, it looks like we should replace `crypto.Signer` with something that has `context.Context` in the signature, that way the sign operation can actually be cancelled after the timeout, not just discarded (but kept running)

Reply:
Yea, I didn't see a way to do this without modifying the implementation of the signer inside the vendored dependencies, but if there was a way to do this outside I'd love to change it.